### PR TITLE
Added test workflow for PR and upgraded dependencies

### PR DESCRIPTION
**Pull request by sonamtenzin6**
## Description
- Enabled build and unit tests in the repo
- Fixed existing tests
- Sorting logic edge case fix in SortIPAddrsByNetworkSize
- Upgraded libraries which could cause this issue - https://hashicorp.atlassian.net/browse/SECVULN-17075

### CODE REVIEW

File: ipaddrs.go [hunk 1]
```diff
@@ -67,9 +67,12 @@ func (s SortIPAddrsByNetworkSize) Less(i, j int) bool {
 	// that have a port (e.g. a host with a /32 and port number is more
 	// specific and should sort first over a host with a /32 but no port
 	// set).
-	if s.IPAddrs[i].IPPort() == 0 || s.IPAddrs[j].IPPort() == 0 {
```
Comment (sonamtenzin6) [inline ipaddrs.go:70]: This condition is currently incorrectly sorting for scenarios like these - 

If we get an order like this, it will return false maintaining that 1>2
1) 128.95.120.2:8600
2) 128.95.120.2
And if we get input like this, it will still return 1>2
1) 128.95.120.2
2) 128.95.120.2:8600

It is essentially not sorting for these scenarios. The test  `TestSockAddr_IPAddrs_IPAddrsByNetworkSize` was failing because of the issue

File: template/template_test.go [hunk 2]
```diff
@@ -25,14 +186,14 @@ func TestSockAddr_Parse(t *testing.T) {
 			output: ``,
 			fail:   true,
 		},
-		{
-			name:   "GetDefaultInterface",
-			input:  `{{GetDefaultInterfaces | include "type" "IPv4" | attr "name" }}`,
-			output: `en0`,
-		},
+		// {
```
Comment (sonamtenzin6) [inline template/template_test.go:189]: Commented out these tests because they are dependent on underlying interfaces of the machine, which may change

File: template/template_test.go [hunk 3]
```diff
@@ -16,7 +177,7 @@ func TestSockAddr_Parse(t *testing.T) {
 	}{
 		{
 			name:   `basic include "name"`,
-			input:  `{{GetAllInterfaces | include "name" "lo0" | printf "%v"}}`,
+			input:  `{{. | include "name" "lo0" | printf "%v"}}`,
```
Comment (sonamtenzin6) [inline template/template_test.go:180]: Changed this from `GetAllInterfaces` because this caused the test to depend on underlying machine interfaces. We are not getting mock input from `getInputList` method

File: template/template_test.go [hunk 4]
```diff
@@ -265,7 +426,7 @@ func TestSockAddr_Parse(t *testing.T) {
 		}
 		t.Run(test.name, func(t *testing.T) {
 			t.Parallel()
-			out, err := socktmpl.Parse(test.input)
+			out, err := socktmpl.ParseIfAddrs(test.input, getInputList())
```
Comment (sonamtenzin6) [inline template/template_test.go:429]: Changed to include mocked interfaces. `Parse` is essentially calling `ParseIfAddrs`

File: .github/workflows/build_test.yml [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    build:
+        runs-on: ubuntu-latest
+
+        steps:
+            - name: Checkout Code
+              uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683
+            - name: Setup Go
+              uses: actions/setup-go@3041bf56c941b39c61721a86cd11f3bb1338122a
+              with:
+                go-version: '1.19'
```
Comment (johanbrandhorst) [inline .github/workflows/build_test.yml:17]: This is an ancient release. I see that the `go.mod` says 1.19, but that just means it's the oldest version supported. We should probably use a matrix here, including 1.19, 1.20, 1.21, 1.22 and 1.23. See https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/running-variations-of-jobs-in-a-workflow#using-a-matrix-strategy if you're unsure how to create a matrix strategy.

Also CC @schmichael, what is this repositories Go compatibility policy? I'd probably suggest we update the `go` directive to 1.22 and use the same policy as the Go team.

File: template/template_test.go (hunk 2 above)
Comment (johanbrandhorst) [reply template/template_test.go:189]: This sounds reasonable to me, but maybe just remove the test?

File: template/template_test.go [hunk 6]
```diff
@@ -167,95 +328,95 @@ func TestSockAddr_Parse(t *testing.T) {
...
-  {{- end -}}
-{{- end -}}`,
-			output: `true`,
-		},
+		//   {{- if ($addr | len) gt 0 -}}
+		//     {{- print "true" -}}{{/* print $addr*/ -}}
+		//   {{- end -}}
+		// {{- end -}}`,
+		// 			output: `true`,
+		// 		},
```
Comment (johanbrandhorst) [inline template/template_test.go:344]: Can this test also not work?

File: template/template_test.go [hunk 7]
```diff
@@ -167,95 +328,95 @@ func TestSockAddr_Parse(t *testing.T) {
...
+		// 	input:  `{{GetPublicIPs}}`,
+		// 	output: `1.2.3.4 6.7.8.9`,
+		// },
+		// {
+		// 	// Assume the private IPs on this host are just the IPv4 addresses:
+		// 	// 10.1.2.3 and 172.16.4.6
+		// 	name:   "GetInterfaceIPs",
+		// 	input:  `{{GetInterfaceIPs "en0"}}`,
+		// 	output: `10.1.2.3 and 172.16.4.6`,
+		// },
```
Comment (johanbrandhorst) [inline template/template_test.go:419]: Remove these

File: template/template_test.go (hunk 4 above)
Comment (johanbrandhorst) [reply template/template_test.go:429]: Instead of a function that gets called in each test, can we not just create this list once at the top of this test function and then use the value in all the test? Does `ParseIfAddrs` mutate the input? If so, could we instead create a copy of the list in each test with `copy`? I think I'd prefer not to have this helper function defined at all.

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]
Thanks for this, the security vulnerability fix is just the mod update, and I'm guessing you decided to fix the tests because they were failing on your machine? I think that makes sense. I added some comments, let me know what you think!

File: .github/workflows/build_test.yml (hunk 5 above)
Comment (schmichael) [reply .github/workflows/build_test.yml:17]: SGTM. This repo doesn't need special treatment. It's just been stable/neglected.

File: template/template_test.go (hunk 6 above)
Comment (schmichael) [reply template/template_test.go:344]: As above I'm +1 for removing ancient tests as opposed to commenting them out.

**Review by schmichael — DISMISSED** [review]
+1 to @johanbrandhorst's comments about bumping Go and dropping commented out tests.

Logic fix looks good.

Can we update the title and description to drop the `security vulnerabilities` reference or at least clarify it's a vulnerability *in a dependency that is unused by go-sockaddr.* go-sockaddr does *not* have a vulnerability, and I would hate people (or security scanners) to flag the wrong package!

Please correct me if I'm missing something.

**Comment by schmichael** [comment]
This will supercede and close #50. Many apologies for missing the opportunity to merge that one!

File: ipaddrs.go (hunk 1 above)
Comment (thevilledev) [reply ipaddrs.go:70]: Just to note here, proposal to fix this has been in #50 since two years ago. It also includes an explanation why this became an issue with Go 1.19 onwards.

Comment (sonamtenzin6) [reply ipaddrs.go:70]: This PR will anyways need some changes to go through, I will merge #50 and take a rebase from master to incorporate the fix

File: template/template_test.go (hunk 4 above)
Comment (sonamtenzin6) [reply template/template_test.go]: Sure, removed the function and replaced with a list

File: .github/workflows/build_test.yml (hunk 5 above)
Comment (sonamtenzin6) [reply .github/workflows/build_test.yml]: Added a matrix for testing on different versions

File: template/template_test.go (hunk 6 above)
Comment (sonamtenzin6) [reply template/template_test.go]: The test is currently failing, so removed it

**Review by schmichael — DISMISSED** [review]

File: ifaddrs_test.go [hunk 8]
```diff
@@ -632,7 +633,7 @@ func TestGetIfAddrs(t *testing.T) {
 		t.Fatalf("No loopback interfaces found, loInt nil")
 	}
 
-	if val := sockaddr.IfAddrAttr(*loInt, "flags"); !(val == "up|loopback|multicast" || val == "up|loopback") {
+	if val := sockaddr.IfAddrAttr(*loInt, "flags"); !strings.Contains(val, "up|loopback") {
```
Comment (johanbrandhorst) [inline ifaddrs_test.go:636]: Why was this changed? This isn't the same behavior anymore.

File: go.mod [hunk 9]
```diff
@@ -27,6 +27,6 @@ require (
 	github.com/posener/complete v1.1.1 // indirect
 	github.com/shopspring/decimal v1.2.0 // indirect
 	github.com/spf13/cast v1.3.1 // indirect
-	golang.org/x/crypto v0.17.0 // indirect
-	golang.org/x/sys v0.22.0 // indirect
+	golang.org/x/crypto v0.32.0 // indirect
```
Comment (johanbrandhorst) [inline go.mod:30]: Could you also change the `go` stanza at the top of the file to `1.23`?

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]

File: ifaddrs_test.go (hunk 8 above)
Comment (sonamtenzin6) [reply ifaddrs_test.go]: We are now getting, `up|loopback|running` as the response on CI and `up|loopback|multicast|running` on local when running with go1.23. Changed the function from contains check to a stricter exact string match

**Review by johanbrandhorst — APPROVED** [review]
Thanks for this, this repo needed a bit of love, just in time for valentines day ❤️

**Review by mohanmanikanta2299 — APPROVED** [review]
LGTM
